ESPController/src/main.cpp

```diff
@@ -1380,6 +1378,168 @@ void enqueue_task(void *param)
+  esp_event_handler_instance_t instance_got_ip;
+  ESP_ERROR_CHECK(esp_event_handler_instance_register(WIFI_EVENT,
+                                                      ESP_EVENT_ANY_ID,
+                                                      &event_handler,
+                                                      NULL,
+                                                      &instance_any_id));
+  ESP_ERROR_CHECK(esp_event_handler_instance_register(IP_EVENT,
+                                                      IP_EVENT_STA_GOT_IP,
+                                                      &event_handler,
+                                                      NULL,
+                                                      &instance_got_ip));
+
+  wifi_config_t wifi_config;
+
+  memset(&wifi_config, 0, sizeof(wifi_config));
+
+  wifi_config.sta.threshold.authmode = WIFI_AUTH_WPA2_PSK;
+  strncpy((char *)wifi_config.sta.ssid, DIYBMSSoftAP::Config()->wifi_ssid, sizeof(wifi_config.sta.ssid));
+  strncpy((char *)wifi_config.sta.password, DIYBMSSoftAP::Config()->wifi_passphrase, sizeof(wifi_config.sta.password));
+
+  ESP_ERROR_CHECK(esp_wifi_set_mode(WIFI_MODE_STA));
+  ESP_ERROR_CHECK(esp_wifi_set_config(WIFI_IF_STA, &wifi_config));
+  ESP_ERROR_CHECK(esp_wifi_start());
+
+  uint32_t chipId = 0;
+  for (int i = 0; i < 17; i = i + 8)
+  {
+    chipId |= ((ESP.getEfuseMac() >> (40 - i)) & 0xff) << i;
+  }
+  // DIYBMS-00000000
+  memset(&hostname,0,sizeof(hostname));
+  snprintf(hostname, sizeof(hostname), "DIYBMS-%08X", chipId);
+  esp_err_t ret = tcpip_adapter_set_hostname(TCPIP_ADAPTER_IF_STA, hostname);
+  if (ret != ESP_OK)
+  {
+    ESP_LOGE(TAG, "failed to set hostname:%d", ret);
+  }
+  ESP_LOGI(TAG, "Hostname: %si", hostname);
+ 
+
+  ESP_LOGI(TAG, "wifi_init_sta finished");
+
+  /* Waiting until either the connection is established (WIFI_CONNECTED_BIT) or connection failed for the maximum
+   * number of re-tries (WIFI_FAIL_BIT). The bits are set by event_handler() (see above) */
+  EventBits_t bits = xEventGroupWaitBits(s_wifi_event_group,
+                                         WIFI_CONNECTED_BIT | WIFI_FAIL_BIT,
+                                         pdFALSE,
+                                         pdFALSE,
+                                         portMAX_DELAY);
+
+  /* xEventGroupWaitBits() returns the bits before the call returned, hence we can test which event actually
+   * happened. */
+  if (bits & WIFI_CONNECTED_BIT)
+  {
+    ESP_LOGI(TAG, "connected to access point");
+  }
+  else if (bits & WIFI_FAIL_BIT)
+  {
+    ESP_LOGI(TAG, "Failed to connect to access point");
+  }
+  else
+  {
+    ESP_LOGE(TAG, "UNEXPECTED EVENT");
+  }
+
+  /* The event will not be processed after unregister */
+  ESP_ERROR_CHECK(esp_event_handler_instance_unregister(IP_EVENT, IP_EVENT_STA_GOT_IP, instance_got_ip));
+  ESP_ERROR_CHECK(esp_event_handler_instance_unregister(WIFI_EVENT, ESP_EVENT_ANY_ID, instance_any_id));
```

Comment:
you will want to keep these objects as there will be disconnect / reconnect events raised periodically.